src/passes/OptimizeInstructions.cpp

```diff
@@ -1904,6 +1920,39 @@ struct OptimizeInstructions
     }
   }
 
+  // Tries pushing the struct.new down so that it is closer
+  // to a potential struct.set.
+  bool trySwap(ExpressionList& list, Index i, Index j) {
+    if (j == list.size() - 1) {
+      // There is no reason to swap with the last element
+      // of the list as it won't match the pattern.
+      return false;
+    }
+
+    if (list[j]->is<LocalSet>() &&
+        list[j]->dynCast<LocalSet>()->value->is<StructNew>()) {
+      // Don't swap two struct.new instructions to avoid going back and forth.
+      return false;
+    }
+    // Check if the local is referencenced by the instruction we want to
+    // swap it with,
+    auto* localSet = list[i]->dynCast<LocalSet>();
+    auto otherEffects = effects(list[j]);
+    if (otherEffects.localsRead.count(localSet->index) ||
+        otherEffects.localsWritten.count(localSet->index)) {
+      return false;
+    }
+    // or if the effects don't permit moving one past the other.
+    auto structNewEffects = effects(localSet->value);
+    if (otherEffects.invalidates(structNewEffects)) {
+      return false;
+    }
```

Comment:
These 10 lines are similar to logic in `optimizeSubsequentStructSet` - I wonder if we can share them?

Reply:
Hmm. I based my code on those 10 lines but I couldn't think of a way to conceptually abstract the logic into a function that had a clear meaning without moving the first check into the loop increasing its cost. I was thinking something like

```
bool canMove(ExpressionList& list, Expression* expr, Index stuctNewLocalIndeex) {
...
}
```

or something like that but I didn't see any utilites to create a sublist and adding yet another index started looking quite obscure.

Reply:
Hmm, yeah, maybe it doesn't make sense to try to combine these.

Actually though I think the localsRead/Written part is not needed here. The `struct.set` is always in a `local.set`, correct? Then `invalidates` will return false if the local is used, since a set of a local cannot be moved past a get.

Reply:
> the struct.set is always in a local.set, correct?
 
There might not be a `struct.set`, just an arbitrary expression we want to swap. And if it is a `struct.set` it is guaranteed to be on a different local. Not that the `local.set` does not have its effects computed but its value is.  My understanding is that a `read` of the local would not invalidate a `local.set`, is that incorrect?

The case would be
```
(local.set $0 (struct.new ...)) ;;1
(....  local.get $0 ...) ;; 2
```

Currently we check that 2 does not read or write `$0` and all its effects don't invalidate anything in the `(struct.new ...)` in 1. If that is the case the expressions can be swapped. But if the `read` or `write` on the local happening in 2 invalidate the effects of the whole expression 1, then yes we could remove the locals check part and just use `invalidate`. 

Shouldn't we also check that 1 invalidates 2? I am imagining that in 1 there could be a local.tee that sets a value to be used in 2 and that should also prevent the swap, but I didn't see that in the code for the current optimization and the concern would be the same.

Reply:
Sorry, I wrote `the struct.set is always in a local.set` but I meant to write `the struct.new is always..`.

We can compute the effects of the entire set of the new, to avoid that second check. That is `effects(localSet->value)` should be `effects(localSet)`. Then we are just checking if two instructions can be swapped, and `invalidate()` checks everything relevant for that.

Reply:
Done. BTW is `.invalidates` symmetrical?

Reply:
Yes, it is symmetrical.